src/values/acceleration-structure.cxx

```diff
@@ -0,0 +1,1484 @@
+        rayFlagSkipAABBs = rayFlags & spv::RayFlagsMask::RayFlagsSkipAABBsKHRMask; // skip procedurals
+    }
+
+    void clearTrace() {
+        while (!nodesToEval.empty()) {
+            nodesToEval.pop();
+        }
+    }
+
+    void initTrace() {
+        closestInstance = nullptr;
+        closestPrimitiveID = 0;
+        committedIntersectionType = CommittedIntersectionType::None;
+        candidateIntersectionType = CandidateIntersectionType::Invalid;
+        activeTrace = true;
+        nodesToEval.push(&root);
+    }
+
+public: // TODO: also traversal stuff
+
+    void resetTrace() {
+        activeTrace = false;
+        clearTrace();
+        initTrace();
+    }
+    
+    void initTrace(const unsigned& rayFlags,
+            const unsigned& cullMask,
+            const std::vector<float>& rayOrigin,
+            const std::vector<float>& rayDirection,            
+            const float& rayTMin,
+            const float& rayTMax,
+            const bool& useSBT,
+            const unsigned offsetSBT = 0,
+            const unsigned strideSBT = 0,
+            const unsigned missIndex = 0) {
+        
+        assert(rayOrigin.size() == 3 && rayDirection.size() == 3);
+        glm::vec4 convertedRayOrigin = glm::make_vec4(rayOrigin.data());
+        convertedRayOrigin.w = 1.0f;
+        glm::vec4 convertedRayDirection = glm::make_vec4(rayDirection.data());
+        convertedRayDirection.w = 0.0f;
+
+        this->rayFlags = rayFlags;
+        this->cullMask = cullMask;
+        this->rayOrigin = convertedRayOrigin;
+        this->rayDirection = convertedRayDirection;
+        this->rayTMin = rayTMin;
+        this->rayTMax = rayTMax;
+
+        this->useSBT = useSBT;
+        this->offsetSBT = offsetSBT;
+        this->strideSBT = strideSBT;
+        this->missIndex = missIndex;
+
+        setFlags();
+        resetTrace();
+    }
+
+    void initTrace(const unsigned& rayFlags,
+            const unsigned& cullMask,
+            const glm::vec4& rayOrigin,
+            const glm::vec4& rayDirection,
+            const float& rayTMin,
+            const float& rayTMax,
+            const bool& useSBT,
+            const unsigned offsetSBT = 0,
+            const unsigned strideSBT = 0,
+            const unsigned missIndex = 0) {
+                
+        this->rayFlags = rayFlags;
+        this->cullMask = cullMask;
+        this->rayOrigin = rayOrigin;
+        this->rayDirection = rayDirection;
+        this->rayTMin = rayTMin;
+        this->rayTMax = rayTMax;
+
+        this->useSBT = useSBT;
+        this->offsetSBT = offsetSBT;
+        this->strideSBT = strideSBT;
+        this->missIndex = missIndex;
+
+        setFlags();
+        resetTrace();
+    }
+
+private:
+    bool didPopNodePreviously = true;
+
+public:
+    // TODO: step to the next primitive
+    // TODO: change instance node pop and push logic
+    bool stepTrace(bool& didIntersectGeometry) {
```

Comment:
I am not generally a fan of including parameters as editable pseudo-returns. Please return a tuple or struct containing both !nodesToEval.empty() and did_intersect_geometry, as needed.